pkg/piecestore/rpc/server/readerwriter.go

```diff
@@ -0,0 +1,41 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package server
+
+import (
+	"bytes"
+
+	pb "storj.io/storj/protos/piecestore"
+)
+
+// StreamWriter -- Struct for writing piece to server upload stream
+type StreamWriter struct {
+	stream pb.PieceStoreRoutes_RetrieveServer
+}
+
+// Write -- Write method for piece upload to stream
+func (s *StreamWriter) Write(b []byte) (int, error) {
+	// Write the buffer to the stream we opened earlier
+	if err := s.stream.Send(&pb.PieceRetrievalStream{Size: int64(len(b)), Content: b}); err != nil {
+		return 0, err
+	}
+
+	return len(b), nil
+}
+
+// StreamReader -- Struct for Retrieving data from server
+type StreamReader struct {
+	stream pb.PieceStoreRoutes_StoreServer
+}
+
+// Read -- Read method for piece download from stream
+func (s *StreamReader) Read(b []byte) (int, error) {
+	storeMsg, err := s.stream.Recv()
+	if err != nil {
+		return 0, err
+	}
+
+	// Write chunk received to disk
+	return bytes.NewReader(storeMsg.Content).Read(b)
```

Comment:
holds up the merge. ah shoot this is so close.

okay so this actually has the exact same problem as before with the overflowData but it's more subtly lost here. let's say `storeMsg.Content` is more data than fits in `b`. what happens? well, we make a `*bytes.Reader` that wraps `storeMsg.Content`. then we call `Read` on that `*bytes.Reader` which will return the beginning of `storeMsg.Content` that fits in `b`. what happens to the rest of `storeMsg.Content`? it gets lost in that temporary `*bytes.Reader` we created just for this `Read` invocation. so we lost some data.

you could make a single bytes.Buffer i suppose, but it's probably easier to do the overflowData thing here again.

this is so close to being able to reuse the code from the previous `StreamReader` it really makes me feel like there's something we should be able to do. if both streams returned the same protobuf on `Recv()`, you could just make one `StreamReader` instead of two that took an interface?

Reply:
they close the streams differently so I kept them separate

Reply:
The only code from the two stream reader/writers that is the same is the reader. the closes, and the sends are different. I want to combine the reads, but not sure the best way to do it (ie folder structure and how to keep it from getting convoluted because it is 2 separate files)

Reply:
The same code will be there twice until I can figure out the best way to do it after we talk about it